### PR TITLE
Ensure that all actions are compared to fixed point in time (ie. world update start)

### LINKED ISSUE #10218: [Hunter] Feign Death + Freezing Arrow bug
using feign death and casting freezing arrow after,causes the target to be trapped instantly.(can use it without using Scattershot)its an annoying pvp bug & Hunters are abusing it.
https://github.com/TrinityCore/TrinityCore/commit/447ddceabad28a3a4ed1ed3bdece632d2d072246
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1412820-hunter-feign-death-freezing-arrow-bug?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #10530: 3.3.5a Hunter Freezing Trap Arming Time
Hunter Freezing Trap Arming time should be 1 second but it's 2 seconds & the range is bugged also.
https://github.com/TrinityCore/TrinityCore/commit/de27cfd10243e266bc2dd275c6d1653258ebd6fb
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/1412908-3-3-5a-hunter-freezing-trap-arming-time?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ghost**
**Target branch(es):** 3.3.5

- [x] 3.3.5

**Issues addressed:** Closes #10530, can potentially fix #10218 if arming time / 2 fix is correct

**Tests performed:** Compiles, simple ingame tests

Ensure that all actions are compared to fixed point in time (eg. world update start) and are not dependent on processing moment. Increased GameObjects cooldown resolution to milliseconds, fixes arming time of traps to be exactly one second and not something from range (1000, 1999)

### CODE REVIEW

File: src/server/game/Spells/SpellHistory.h [hunk 1]
```diff
@@ -21,6 +21,7 @@
 #include "SharedDefines.h"
 #include "QueryResult.h"
 #include "Transaction.h"
+#include "World.h"
```
Comment (Shauren) [inline src/server/game/Spells/SpellHistory.h]: pls no.

Comment (xinef1) [inline src/server/game/Spells/SpellHistory.h]: Separate static instance containing all gametime related stuff?

File: src/server/game/Time/UpdateTime.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+UpdateTime::UpdateTime()
+{
+    _averageUpdateTime = 0;
+    _totalUpdateTime = 0;
+    _updateTimeTableIndex = 0;
+    _maxUpdateTime = 0;
+    _maxUpdateTimeOfLastTable = 0;
+    _maxUpdateTimeOfCurrentTable = 0;
+
+    _updateTimeDataTable.resize(AVG_DIFF_COUNT, 0);
```
Comment (Shauren) [inline src/server/game/Time/UpdateTime.cpp]: is this the only resize? use std::array then

Comment (xinef1) [inline src/server/game/Time/UpdateTime.cpp]: hmm completely forgot about it, used vector because i wanted to avoid use of pure table

**Review by Naios — APPROVED** [review]

File: src/server/game/Entities/GameObject/GameObject.h [hunk 3]
```diff
@@ -516,7 +516,7 @@ struct GameObjectTemplate
             case GAMEOBJECT_TYPE_AREADAMAGE:    autoCloseTime = areadamage.autoCloseTime; break;
             default: break;
         }
-        return autoCloseTime / IN_MILLISECONDS;              // prior to 3.0.3, conversion was / 0x10000;
```
Comment (Naios) [inline src/server/game/Entities/GameObject/GameObject.h]: Why this is removed?

File: src/server/game/Time/GameTime.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+        /// Uptime (in secs)
+        uint32 GetUptime() const;
+
+        void UpdateGameTimers();
+        void SetStartTime();
+
+    private:
+        // Game Time
+        time_t _startTime;
```
Comment (Naios) [inline src/server/game/Time/GameTime.h]: Initializing the fields here allows you to define the constructor as default -> `GameTime() = default;` which is much better style.

File: src/server/game/World/World.h [hunk 5]
```diff
@@ -631,16 +629,6 @@ class TC_GAME_API World
...
 
-        /// When server started?
-        time_t const& GetStartTime() const { return m_startTime; }
-        /// What time is it?
-        time_t const& GetGameTime() const { return m_gameTime; }
-        /// Uptime (in secs)
-        uint32 GetUptime() const { return uint32(m_gameTime - m_startTime); }
-        /// Update time
-        uint32 GetUpdateTime() const { return m_updateTime; }
-        void SetRecordDiffInterval(int32 t) { if (t >= 0) m_int_configs[CONFIG_INTERVAL_LOG_UPDATE] = (uint32)t; }
```
Comment (Naios) [inline src/server/game/World/World.h]: Also remove the removed config keys from the config please.

**Review by Naios — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/GameObject/GameObject.h (hunk 3 above)
Comment (Shauren) [inline src/server/game/Entities/GameObject/GameObject.h]: look at gameobject.cpp - timer is stored in ms now

File: src/server/game/World/World.h (hunk 5 above)
Comment (xinef1) [inline src/server/game/World/World.h]: https://github.com/TrinityCore/TrinityCore/pull/18910/files#diff-2094dfd21434b2fec5f1ba58018139f7R110 moved here

File: src/server/game/Time/GameTime.h [hunk 6]
```diff
@@ -53,11 +53,11 @@ class TC_GAME_API GameTime
 
     private:
         // Game Time
-        time_t _startTime;
-        time_t _gameTime;
-        uint32 _gameMSTime;
-        std::chrono::system_clock::time_point _gameTimeSystemPoint;
-        std::chrono::steady_clock::time_point _gameTimeSteadyPoint;
+        time_t _startTime = time(nullptr);
```
Comment (Shauren) [inline src/server/game/Time/GameTime.h]: tbh I'm not sure I agree with @Naios here embedding library function calls in class body to initialize members...

Comment (xinef1) [inline src/server/game/Time/GameTime.h]: I agree with Shauren, but i'm only a peasant. Call me if you come to an agreement.

Comment (Naios) [inline src/server/game/Time/GameTime.h]: Ah no, you are right, that should stay in the compilation unit.

**Review by Naios — APPROVED** [review]

**Pull request merged**